### ISSUE TITLE
False positive in `Lint/UselessAccessModifier` when using `private` in singleton method

### ISSUE THREAD

**Issue by eiskrenkov** [rubocop/rubocop#14334]
`Lint/UselessAccessModifier` is getting confused when calling `private` in singleton method and then using `private` again on an instance level

<details>
  <summary>Code example</summary>
  
  ```ruby
  module A
    extend ActiveSupport::Concern
  
    # Just an example implementation of class method. Real issue was encountered when using
    # ActiveSupport::Concern.class_methods
    #
    def self.test(&)
      instance_eval(&)
    end
  
    test do
      def a
        puts "a"
      end
  
      private
  
      def b
        puts "b"
      end
    end
  
    def a
      puts "a"
    end
  
    private # <- Troublesome line
  
    def b
      puts "b"
    end
  end
 
  # my-project(dev)* module A
  # my-project(dev)*   extend ActiveSupport::Concern
  # my-project(dev)*
  # my-project(dev)*   def self.test(&)
  # my-project(dev)*     instance_eval(&)
  # my-project(dev)*   end
  # my-project(dev)*
  # my-project(dev)*   test do
  # my-project(dev)*     def a
  # my-project(dev)*       puts "a"
  # my-project(dev)*     end
  # my-project(dev)*
  # my-project(dev)*     private
  # my-project(dev)*
  # my-project(dev)*     def b
  # my-project(dev)*       puts "b"
  # my-project(dev)*     end
  # my-project(dev)*   end
  # my-project(dev)*
  # my-project(dev)*   def a
  # my-project(dev)*     puts "a"
  # my-project(dev)*   end
  # my-project(dev)*
  # my-project(dev)*   private
  # my-project(dev)*
  # my-project(dev)*   def b
  # my-project(dev)*     puts "b"
  # my-project(dev)*   end
  # my-project(dev)> end
  # => :b
  # my-project(dev)> A.a
  # a
  # => nil
  # my-project(dev)> A.b
  # (my-project):102:in '<main>': private method 'b' called for module A (NoMethodError)
  # my-project(dev)> class B; include A; end
  # => B
  # my-project(dev)> B.new.a
  # a
  # => nil
  # my-project(dev)> B.new.b
  # (my-project):105:in '<main>': private method 'b' called for an instance of B (NoMethodError)
  ```
</details>

--------

## Expected behavior

Rubocop doesn't mark instance level `private` modifier as useless

## Actual behavior

```sh
~/t/my-project ❯ rubocop /Users/eiskrenkov/tds/my-project/lib/test.rb --debug
Using Gemfile's rubocop
For /Users/eiskrenkov/tds/my-project: configuration from /Users/eiskrenkov/tds/my-project/.rubocop.yml
Inheriting configuration from /Users/eiskrenkov/tds/my-project/.rubocop-shared.yml
rubocop-rspec extension supports plugin, specify `plugins: rubocop-rspec` instead of `require: rubocop-rspec` in /Users/eiskrenkov/tds/my-project/.rubocop-shared.yml.
For more information, see https://docs.rubocop.org/rubocop/plugin_migration_guide.html.

Default configuration from /Users/eiskrenkov/.local/share/mise/installs/ruby/3.4.4/lib/ruby/gems/3.4.0/gems/rubocop-1.77.0/config/default.yml
Plugin configuration from /Users/eiskrenkov/.local/share/mise/installs/ruby/3.4.4/lib/ruby/gems/3.4.0/gems/rubocop-rspec-3.6.0/config/default.yml
Inheriting configuration from /Users/eiskrenkov/tds/my-project/.rubocop-base.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/eiskrenkov/tds/my-project/lib/test.rb
Loading cache from /Users/eiskrenkov/.cache/rubocop_cache/180a51ef8862870ddfbbefaa9784da3a5c96d804/6d7a3b621ca1730e04accd938619e4bdab66cfb1/eec6529c838ebbea6f2b646671e39bd363bc717b
W

Offenses:

lib/test.rb:29:3: W: [Correctable] Lint/UselessAccessModifier: Useless private access modifier.
  private
  ^^^^^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable
Finished in 0.23777699982747436 seconds
```

## Steps to reproduce the problem

Call `private` inside of block passed to singleton method and then call `private` on instance level. Example:

Paste

<details>
  <summary>Code example</summary>
  
  ```ruby
  module A
    extend ActiveSupport::Concern
  
    # Just an example implementation of class method. Real issue was encountered when using
    # ActiveSupport::Concern.class_methods
    #
    def self.test(&)
      instance_eval(&)
    end
  
    test do
      def a
        puts "a"
      end
  
      private
  
      def b
        puts "b"
      end
    end
  
    def a
      puts "a"
    end
  
    private # <- Troublesome line
  
    def b
      puts "b"
    end
  end
  ```
</details>

into `test.rb` and run `rubocop test.rb`

## RuboCop version

```
$ bundle exec rubocop -V
1.77.0 (using Parser 3.3.8.0, Prism 1.4.0, rubocop-ast 1.45.1, analyzing as Ruby 3.4, running on ruby 3.4.4) [arm64-darwin24]
  - rubocop-rspec 3.6.0
```

**Comment by Earlopain** [comment]
You can configure this with `ContextCreatingMethods`. If you use the latest `rubocop-rails`, it will automatically do so for all the `Concern` methods like `class_methods` and `included` etc. Otherwise, you will need to add these to your config yourself.

https://github.com/rubocop/rubocop-rails/blob/242b6b40e801e257fae3dc5aa06edec1b52a2d80/config/default.yml#L80-L92

**Comment by eiskrenkov** [comment]
Got it, @Earlopain, thanks for your help! :D
